### DISCUSSION TITLE
Search

### DISCUSSION THREAD

**Discussion by moloch--** [vriteio/vrite#14] [Ideas]
Search functionality in the web UI and API

**Comment by areknawo** [comment]
There are 2 types of "search" I have planned:
- find & replace right in the editor (inspired by similar features in code editors like VS Code, with support for regex, operating only on content available in the editor;
- general search (2 stages):
  1. initially querying only data like description, title, and other metadata, together with a command palette for faster navigation in the entire app;
  2. full-text/vector search for querying through the entire available dataset (metadata & content); it's the end goal for a great search, but it will take the most time to implement;

Let me know which type of search you were referring to, so I can prioritize its development.

**Reply by moloch--** [reply]
General Search / (ii) full text is the biggest thing missing for what we'd like to use vrite for, with (i) being a close second.

**Reply by areknawo** [reply]
Makes sense. (i) is coming in the next week or two, as it's simpler to implement. (ii) will come later on (likely late July/early August).

**Reply by areknawo** [reply]
@moloch-- Took a little longer than planned, but a search is now available in **v2023.8.18-beta** with the addition of a command palette. You can both search through your content and ask questions (powered by hybrid vector search and GPT-3.5). `search()` method is already available in the SDK with the same capabilities as the built-in search, while additional `/search/ask` endpoint is also available with SSE streaming. Docs for both and SDK support for the `/search/ask` endpoint will be added soon.

**Comment by saiqulhaq** [comment]
![image](https://github.com/vriteio/vrite/assets/1275215/efde8079-3420-4f20-8354-820e4147416b)
I am using https://app.vrite.io/
I am having an issue to add my OpenAI access token
Could you share what is the cause?
Should I use self-hosted version to get make OpenAI extension works?

**Reply by areknawo** [reply]
In the hosted version you don't supply the token.

The AI/semantic search is built-in and enabled by default. You use the search bar in the dashboard or a `⌘K`/`Ctrl K` shortcut to open the palette.

![search](https://github.com/vriteio/vrite/assets/10221873/b93c0c5f-6beb-448b-9d08-7a3115096869)

On the other hand, the GPT-3.5 extension integrates the AI into the editor. Once installed, you can use the new "Block Action" button (on the right side of the editor) to use GPT-3.5 to generate content right into Vrite Editor.

![block-action](https://github.com/vriteio/vrite/assets/10221873/f812847b-0803-4c37-bd6f-dbc21a833c6a)

If you want to self-host then you can choose to enable the AI/semantic search by providing your OpenAI token as an [env variable](https://docs.vrite.io/self-hosting/configuration/).

However, as extensions aren't currently supported while self-hosting, the GPT-3.5 editor integration is only available in the "Vrite Cloud".

**Reply by saiqulhaq** [reply]
Oh, got it
I found two issues:
1. chatgpt "block action" button is missing after I click the "generated button". I think I have to wait a few minutes, then it will appear. Any reasons for this? is there a rate limiter?
2. the extension doesn't read the article I wrote in the editor. so it has no information about what I added to the editor. the chatgpt is it expected to be used for generating articles using their own DB only? can't be used for Question Answer? FYI I have 1 article only

**Reply by areknawo** [reply]
1. This might be a bug. I will open an issue for that and work on fixing it.
2. Right now, the "block action only uses the GPT knowledge and optionally the active paragraph (when *Include context* is used) to **create new content**.

If you want to generate an answer from your existing content base, you can use the *Ask* option in the command palette. Here's both the search and AI "Q&A" in use:

![](https://res.cloudinary.com/practicaldev/image/fetch/s--ibibpyWe--/c_limit%2Cf_auto%2Cfl_progressive%2Cq_66%2Cw_800/https://assets.vrite.io/6409e82d7dfc74cef7a72e0d/EiaM_RSqZBHHEGG5W2TLQ.gif)

So, in general:
- Currently, the extension is for generating new content (e.g. "Write an introduction to JavaScript") or editing paragraphs (e.g. "Make it shorter")
- The semantic search, AI-based question-answering (for your content base), and a command menu are all in the command palette and accessible via API.

**Reply by saiqulhaq** [reply]
Thanks for the answer
I can use "Ask" feature without any issues
By the way, what is the rate limit of "Ask" feature?

**Reply by areknawo** [reply]
Currently, for the API, it's 500 per 1 min.
